### PR TITLE
handle var type inference

### PR DESCRIPTION
**Pull request by panyx0718**

### CODE REVIEW

**Comment by panyx0718** [comment]
@seiriosPlus you can remove the default fill_constant VarTypeInference now

File: python/paddle/fluid/layer_helper.py [hunk 1]
```diff
@@ -324,10 +324,19 @@ def get_parameter(self, name):
...
+        Note:
+            The default type will be set to LOD_TENSOR. However, when
+            the var is used as operator output, its type will be updated
+            based on operator's `VarTypeInference` implementation in
+            infer_var_type.
+        """
         return self.main_program.current_block().create_var(
             name=unique_name.generate(".".join([self.name, 'tmp'])),
             dtype=dtype,
+            type=core.VarDesc.VarType.LOD_TENSOR,
```
Comment (chengduoZH) [inline python/paddle/fluid/layer_helper.py:339]: If the `type` is not set, `Variable` will set the type as `core.VarDesc.VarType.LOD_TENSOR`, the code is [here](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/fluid/framework.py#L257).

**Review by chengduoZH — COMMENTED** [review]
It seems that you mainly replace `create_tmp_variable` with `create_variable_for_type_inference`, what's the different?

File: paddle/fluid/framework/op_desc.cc [hunk 2]
```diff
@@ -515,20 +515,14 @@ void OpDesc::InferShape(const BlockDesc &block) const {
...
-    VLOG(10) << this->Type()
-             << " has not registered InferVarType. Set output variables to "
-                "LOD_TENSOR";
-    for (auto &out_pair : this->outputs_) {
-      for (auto &out_var_name : out_pair.second) {
-        block->FindRecursiveOrCreateVar(out_var_name)
-            .SetType(proto::VarType::LOD_TENSOR);
-      }
-    }
   }
```
Comment (chengduoZH) [inline paddle/fluid/framework/op_desc.cc:526]: I think that every operator should have a `InferVarType`, if not, we should throw exception.

File: python/paddle/fluid/layer_helper.py (hunk 1 above)
Comment (panyx0718) [reply python/paddle/fluid/layer_helper.py:339]: See OpDesc::InferVarType

File: paddle/fluid/framework/op_desc.cc (hunk 2 above)
Comment (panyx0718) [reply paddle/fluid/framework/op_desc.cc:526]: Right...

**Review by chengduoZH — APPROVED** [review]
LGTM

**Pull request merged**